### ISSUE TITLE
issue when item is dnd-draggable and dnd-droppable

### ISSUE THREAD

**Issue by joachim-bl** [akserg/ng2-dnd#105]
* **I'm submitting a ...**
[x ] bug report
[ ] feature request
[ ] question about the decisions made in the repository

* **Do you want to request a *feature* or report a *bug*?**

* **What is the current behavior?**
When i start dragging an item it is instantiated as "DroppableComponent" instead of "DraggableComponent"

* **If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem** via https://plnkr.co or similar.
Add "dnd-draggable" and "dnd-droppable" to the same item

* **What is the expected behavior?**

* **What is the motivation / use case for changing the behavior?**
I try to use it for a tree control to drag folders. Every folder is a draggable and also a drop target. How can I achieve that?

* **Please tell us about your environment:**

- Angular version: 2.2.3
- Browser: [all ]

* **Other information** (e.g. detailed explanation, stacktraces, related issues, suggestions how to fix, links for us to have context, eg. stackoverflow, gitter, etc)

I believe the problem is the abstract component construction. When I look in the debugger, I see that when I start dragging an item the instance of the component is acutally a **DroppableComponent** but since they have the same abstract class, it calls the _onDragStart in the abstract class which will not find the derived method of _onDragStartCallback in the DraggableComponent. Actually this one is never called.
From an architectural point it does not make sense to me to put functionality into a base class (e.g. dragStart) that is only valid for ONE decendant (DraggableComponent).

**Comment by joachim-bl** [comment]
I solved it. I just wrapped the DRAG-EVENTS with a if and set them only for DraggableComponent and the DROP-EVENTS only for the DroppableComponent....and it works! :-)

I am sure there is a cleaner way to do this, though.

`        //
        // DROP events
        //
        if(this.constructor.name == "DroppableComponent"){
            this._elem.ondragenter = (event: Event) => {
                this._onDragEnter(event);
            };
            this._elem.ondragover = (event: DragEvent) => {
                this._onDragOver(event);
                //
                if (event.dataTransfer != null) {
                    event.dataTransfer.dropEffect = this._config.dropEffect.name;
                }

                return false;
            };
            this._elem.ondragleave = (event: Event) => {
                this._onDragLeave(event);
            };
            this._elem.ondrop = (event: Event) => {
                this._onDrop(event);
            };
        }
        //
        // Drag events
        //
        if(this.constructor.name == "DraggableComponent"){
            this._elem.ondragstart = (event: DragEvent) => {
                // console.log('ondragstart', event.target);
                this._onDragStart(event);
                //
                if (event.dataTransfer != null) {
                    event.dataTransfer.setData('text', '');
                    // Change drag effect
                    event.dataTransfer.effectAllowed = this.effectAllowed || this._config.dragEffect.name;
                    // Change drag image
                    if (isPresent(this.dragImage)) {
                        if (isString(this.dragImage)) {
                            (<any>event.dataTransfer).setDragImage(createImage(<string>this.dragImage));
                        } else if (isFunction(this.dragImage)) {
                            (<any>event.dataTransfer).setDragImage(callFun(<Function>this.dragImage));
                        } else {
                            let img: DragImage = <DragImage>this.dragImage;
                            (<any>event.dataTransfer).setDragImage(img.imageElement, img.x_offset, img.y_offset);
                        }
                    } else if (isPresent(this._config.dragImage)) {
                        let dragImage: DragImage = this._config.dragImage;
                        (<any>event.dataTransfer).setDragImage(dragImage.imageElement, dragImage.x_offset, dragImage.y_offset);
                    } else if (this.cloneItem) {
                        this._dragHelper = <HTMLElement>this._elem.cloneNode(true);
                        this._dragHelper.classList.add('dnd-drag-item');
                        this._dragHelper.style.position = "absolute";
                        this._dragHelper.style.top = "0px";
                        this._dragHelper.style.left = "-1000px";
                        this._elem.parentElement.appendChild(this._dragHelper);
                        (<any>event.dataTransfer).setDragImage(this._dragHelper, event.offsetX, event.offsetY);
                    }
                    // Change drag cursor
                    if (this._dragEnabled) {
                        this._elem.style.cursor = this.effectCursor ? this.effectCursor : this._config.dragCursor;
                    } else {
                        this._elem.style.cursor = this._defaultCursor;
                    }
                }
            };
            this._elem.ondragend = (event: Event) => {
                if (this._elem.parentElement && this._dragHelper) {
                    this._elem.parentElement.removeChild(this._dragHelper);
                }
                // console.log('ondragend', event.target);
                this._onDragEnd(event);
                // Restore style of dragged element
                this._elem.style.cursor = this._defaultCursor;
            };
        }
`

**Comment by savvaoff** [comment]
@joachim-bl great! have you created pull request with that fix? 
please, do. really need the same feature to work.

**Comment by joachim-bl** [comment]
My solution is a rather dirty hack, i do not consider it a fix. For a fix the component should be refactored. If you need the fix urgently, just check out the code and add that single if-else parenthesis

**Comment by ChrisTarasovs** [comment]
@joachim-bl  Hi, i am little bit new to node modules and I am unclear on how you applied the fix? Did you apply the fix directly in the node_module, or did you download the git, applied the fix, run the compile and pushed it to npmjs so than you could use in the application? Could you also specific if possible what file u did the update in?

**Comment by vstrizhenok** [comment]
Hi, I proposed solution to use `droppable` and `draggable` for one element like this

<div dnd-draggable [dragData]="entity" [dropZones]="someDropZones"
  dnd-droppable (onDropSuccess)="someSuccessfulHandler($event.dragData)">
....
</div>

**Comment by Omikhleia** [comment]
@vstrizhenok dropZones could be different for the draggable element (i.e; where it can be dropped) and the droppable element (i.e. what can be dropped on it). How is that taken into account?

**Comment by vstrizhenok** [comment]
@Omikhleia Yep but for now dropZones parameter is the same for droppable and draggable elements. For separate using it needs to add new parameter or just wrap on of the component (for example droppable component with own parameter). In my view droppable component must use with dropZones parameter and draggable component must use with dragZones parameter
